src/server/pfs/drive/driver.go

```diff
@@ -217,25 +220,44 @@ func (d *driver) StartCommit(repo *pfs.Repo, commitID string, parentID string, b
 			return err
 		}
 	}
+	d.commitConds[commitID] = sync.NewCond(&d.lock)
 	return nil
 }
 
-func (d *driver) FinishCommit(commit *pfs.Commit, finished *google_protobuf.Timestamp, shards map[uint64]bool) error {
+// FinishCommit blocks until its parent has been finished/cancelled
+func (d *driver) FinishCommit(commit *pfs.Commit, finished *google_protobuf.Timestamp, cancelled bool, shards map[uint64]bool) error {
+	canonicalCommit, err := d.canonicalCommit(commit)
+	if err != nil {
+		return err
+	}
 	// closure so we can defer Unlock
 	var diffInfos []*pfs.DiffInfo
 	if err := func() error {
 		d.lock.Lock()
 		defer d.lock.Unlock()
-		canonicalCommit, err := d.canonicalCommit(commit)
-		if err != nil {
-			return err
-		}
 		for shard := range shards {
 			diffInfo, ok := d.diffs.get(pfsclient.NewDiff(canonicalCommit.Repo.Name, canonicalCommit.ID, shard))
 			if !ok {
 				return fmt.Errorf("commit %s/%s not found", canonicalCommit.Repo.Name, canonicalCommit.ID)
 			}
+			if diffInfo.ParentCommit != nil {
+				parentDiffInfo, ok := d.diffs.get(pfsclient.NewDiff(canonicalCommit.Repo.Name, diffInfo.ParentCommit.ID, shard))
+				if !ok {
+					return fmt.Errorf("parent commit %s/%s not found", canonicalCommit.Repo.Name, diffInfo.ParentCommit.ID)
+				}
+				// Wait for parent to finish
+				for parentDiffInfo.Finished == nil {
+					cond, ok := d.commitConds[diffInfo.ParentCommit.ID]
+					if !ok {
+						return fmt.Errorf("parent commit %s/%s was not finished but a corresponding conditional variable could not be found; this is likely a bug")
+					}
+					cond.Wait()
+				}
+
+				diffInfo.Cancelled = parentDiffInfo.Cancelled
+			}
 			diffInfo.Finished = finished
+			diffInfo.Cancelled = diffInfo.Cancelled || cancelled
```

Comment:
Yeah I'm a fool, `|=` is for bitwise or, it doesn't work on logical or.

Kindly ignore this :p